h2o-algos/src/main/java/hex/gam/GamSplines/ThinPlateRegressionUtils.java

```diff
@@ -0,0 +1,347 @@
+    int maxVal = maxValue(oneBasis);
+    int[] mapFreq = new int[maxVal+1];
+    for (int val : oneBasis)
+      mapFreq[val]++;
+    return mapFreq;
+  }
+  
+  public static void mergeCombos(ArrayList<int[]> oneCombo, Integer[] basisOrder, int[] polyBasis, List<Integer[]> polyBasisSet) {
+    for (int[] oneList : oneCombo) {
+      Arrays.fill(basisOrder, 0);
+      expandCombo(oneList, polyBasis, basisOrder);
+      polyBasisSet.add(basisOrder.clone());
+    }
+  }
+
+  /***
+   * Given a combo found by findOnePerm say for d = 5, m = 4, for degree = 1 to m-1 (3 in this case).  The basis poly
+   * is {3,2,1}.  The returned list of combo are: {{0, 0, 1}, {0, 1, 0}, {0, 0, 2}, {1,0,0}, {0,1,1}, {0,0,3}}.
+   * However, we need to convert this list back to the perspective of the predictors.  In this case, we have 5 
+   * predictors.  This function will translate the list from findOnePerm to the perspective of the predictors.  For 
+   * instance, for degree = 1, we need to have one predictive to have degree of 1 and hence the list should be 
+   * {1, 0, 0, 0, 0}.  For degree = 2, we can have one predictor taking degree of 2 or two predictors each taking 
+   * degree 1.  The same applies to the other degrees.  Hence, this function will return the following list: 
+   * {{1, 0, 0, 0, 0}, {0, 2, 0, 0, 0}, {1, 1, 0, 0, 0}, {0, 0, 3, 0, 0}, {1, 2, 0, 0, 0} and {1, 1, 1, 0, 0}}.
+   * @param oneList
+   * @param polyBasis
+   * @param basisOrder
+   */
+  public static void expandCombo(int[] oneList, int[] polyBasis, Integer[] basisOrder) {
+    int expandIndex = 0;
+    for (int index = 0; index < polyBasis.length; index++) {
+      int count = 0;
+      if (oneList[index] == 0) {
+        basisOrder[expandIndex++] = 0;
+      } else {
+        while (count < oneList[index]) {
+          basisOrder[expandIndex++] = polyBasis[index];
+          count++;
+        }
+      }
+    }
+  }
+
+  /***
+   * For a fixed degree specified as totDegree, specified a set of combination of polynomials to achieve the totDegree.
+   * For instance, if degreeCombo = {1,2,3} and the totDegree is 1. There is only one way to achieve it by the array
+   * {1,0,0}.  If totDegree = 2, there are two arrays that will work: {0,1,0} or {2,0,0}.  If totDegree = 3, there will
+   * be 3 arrays that will work {3,0,0}, {1,1,0}, {0,0,1}.
+   * 
+   * @param totDegree : integer representing degree of polynomial basis
+   * @param degreeCombo : degrees allowed for polynomial basis
+   * @param index
+   * @param allCombos
+   * @param currCombo
+   */
+  public static void findOnePerm(int totDegree, int[] degreeCombo, int index, ArrayList<int[]> allCombos,
+                                 int[] currCombo) {
+    if (totDegree == 0) {
+      if (currCombo != null)
+        allCombos.add(currCombo.clone());
+      return;
+    } else if (totDegree < 0 || index >= degreeCombo.length) { // last added value too big
+      return;
```

Comment:
redundant if-else branch, condition could be modified and merged with else branch

Reply:
Done.  Thanks.